Address safer CPP warnings in PrintStream.h

cdumez opened the pull request:
#### 4754a94673a1728d8c6939c033e8dc0745444cda
<pre>
Address safer CPP warnings in PrintStream.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=302376">https://bugs.webkit.org/show_bug.cgi?id=302376</a>

Reviewed by Keith Miller and Geoffrey Garen.

Address safer CPP warnings in PrintStream.h about StringImpl and
StringView being forward declared. I had to move some functions to other
headers to address circular include problems when trying to include
StringView.h and StringImpl.h in PrintStream.h.

* Source/WTF/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WTF/wtf/PrintStream.h:
(WTF::printInternal):
(WTF::ScopedEnumDump::dump const):
(WTF::EnumDumpWithDefault::dump const):
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::printInternal):
* Source/WTF/wtf/text/UniquedStringImpl.h:
(WTF::printInternal):

Canonical link: <a href="https://commits.webkit.org/303008@main">https://commits.webkit.org/303008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/539e9e6cfd80dea8ef2194232ac8c9d7e699950e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138209 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82430 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99645 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7391d379-ab33-4084-800e-271164294d6d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117106 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80346 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2142 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35235 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81461 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122794 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110733 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35739 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140685 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129236 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2573 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108157 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2896 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108084 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27524 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Keith Miller and Geoffrey Garen; Running compile-webkit") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2181 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31865 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55852 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20380 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2918 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66310 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162251 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2739 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40473 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2939 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2846 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/PrintStreamEnum.h:
```diff
@@ -0,0 +1,80 @@
+void printInternal(PrintStream& out, Enum e)
+{
+    out.print(StringView(enumName(e)));
+}
```
It would probably be better to just have these not rely on a StringView rather than add this header. Something like: 

```
template<typename Enum>
requires (std::is_enum_v<std::decay_t<Enum>>)
void printInternal(PrintStream& out, Enum e)
{
    out.print(spanReinterpretCast<const char8_t>(enumName(e)));
}
```

We know these are ASCII since C++ only supports ASCII source. Alternatively, we could add a `void printInternal(PrintStream&, std::span<const char>);` and use that instead.

 Adding this header is going to cause a lot of headaches as people struggle to understand why they can't print an enum when some other code can.

cdumez replied:
Ok, I'll look into this. Thanks for the suggestion. I agree it would be good not to have a new header if we can avoid it.

cdumez replied:
I updated the PR.

iidmsa commented on Source/WTF/wtf/PrintStreamEnum.h:
```diff
@@ -0,0 +1,80 @@
+namespace WTF {
+
+template<typename Enum>
+requires (std::is_enum_v<std::decay_t<Enum>>)
```
Nit: `requires` clause should be indented
```suggestion
template<typename Enum>
    requires (std::is_enum_v<std::decay_t<Enum>>)
```

kmiller68 commented on Source/WTF/wtf/PrintStream.h:
```diff
@@ -171,7 +165,7 @@ template<typename Enum>
 void printInternal(PrintStream& out, Enum e)
 {
-    out.print(StringView(enumName(e)));
+    out.print(enumName(e));
```
Oh there is a `span<const char>` printer. Nice!

kmiller68 reviewed (approved):
r=me

cdumez committed:
Address safer CPP warnings in PrintStream.h
https://bugs.webkit.org/show_bug.cgi?id=302376

Reviewed by Keith Miller and Geoffrey Garen.

Address safer CPP warnings in PrintStream.h about StringImpl and
StringView being forward declared. I had to move some functions to other
headers to address circular include problems when trying to include
StringView.h and StringImpl.h in PrintStream.h.

* Source/WTF/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WTF/wtf/PrintStream.h:
(WTF::printInternal):
(WTF::ScopedEnumDump::dump const):
(WTF::EnumDumpWithDefault::dump const):
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::printInternal):
* Source/WTF/wtf/text/UniquedStringImpl.h:
(WTF::printInternal):

Canonical link: https://commits.webkit.org/303008@main
